internal/expr/prepare.go

```diff
@@ -56,28 +56,139 @@ func starCount(fns []fullName) int {
 }
 
 // prepareInput checks that the input expression corresponds to a known type.
-func prepareInput(ti typeNameToInfo, p *inputPart) (typeMember, error) {
-	info, ok := ti[p.sourceType.prefix]
-	if !ok {
-		ts := getKeys(ti)
-		if len(ts) == 0 {
-			return nil, fmt.Errorf(`type %q not passed as a parameter`, p.sourceType.prefix)
-		} else {
-			return nil, fmt.Errorf(`type %q not passed as a parameter, have: %s`, p.sourceType.prefix, strings.Join(ts, ", "))
+func prepareInput(ti typeNameToInfo, p *inputPart) ([]fullName, []typeMember, error) {
+	var inCols = make([]fullName, 0)
+	var typeMembers = make([]typeMember, 0)
+
+	numTypes := len(p.sourceTypes)
+	numColumns := len(p.targetColumns)
+	starTypes := starCount(p.sourceTypes)
+	starColumns := starCount(p.targetColumns)
```

Comment:
Would it be better to declare these closer to where they're first used?

Reply:
Moved.